File: ios/LaneShadow/Views/Molecules/MinimalOverlayWidget.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Overlay Type

/**
 * Overlay type enum
 *
 * Defines the three weather overlay types available in the map view.
 */
public enum LSOverlayType: String, CaseIterable, Sendable {
    case wind
    case rain
    case temperature

    /// SF Symbol name for this overlay type
    var icon: String {
        switch self {
        case .wind: "wind"
        case .rain: "drop"
        case .temperature: "thermometer"
        }
    }

    /// Accessibility label for this overlay type
    var accessibilityLabel: String {
        switch self {
        case .wind: "Wind"
        case .rain: "Rain"
        case .temperature: "Temperature"
        }
    }
}

// MARK: - Overlay Availability

/**
 * Overlay availability model
 *
 * Defines which overlay types are currently available for selection.
 */
public struct LSOverlayAvailability: Sendable {
    public let wind: Bool
    public let rain: Bool
    public let temperature: Bool

    public init(
        wind: Bool,
        rain: Bool,
        temperature: Bool
    ) {
        self.wind = wind
        self.rain = rain
        self.temperature = temperature
    }

    /// Returns availability for a given overlay type
    func isAvailable(_ overlayType: LSOverlayType) -> Bool {
        switch overlayType {
        case .wind: wind
        case .rain: rain
        case .temperature: temperature
        }
    }
}

// MARK: - Minimal Overlay Widget Component

/**
 * Minimal overlay widget molecule component
 *
 * A compact, single-icon weather overlay control that expands into a radial menu.
 * Inspired by motorcycle instrument dials and compass navigation.
 * Following React Native component from react-native/components/map/minimal-overlay-widget.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Center button background: `theme.colors.surfaceVariant.default`
 *   - Active indicator border: `theme.colors.primary.default`
 *   - Selected overlay background: `theme.colors.primary.default` at 20% opacity
 *   - Border: `theme.colors.border.default` (1.5pt for center, 1pt for radial icons)
 *   - Unavailable opacity: 0.4
 * - Layout:
 *   - Container: 120×120pt
 *   - Center button: 40×40pt circle
 *   - Active ring: 48×48pt circle
 *   - Radial icons: 36×36pt circles
 *   - Radial radius: 36pt from center
 *   - Center button border: 1.5pt
 *   - Radial icon border: 1pt
 * - Animation:
 *   - Expand/collapse: spring animation
 *   - Icon positions: easeInOut 0.2s
 * - Icons:
 *   - Wind: "wind" (SF Symbol)
 *   - Rain: "drop" (SF Symbol)
 *   - Temperature: "thermometer" (SF Symbol)
 *   - Default (none selected): "square.stack.3d.up"
 *   - Radial icon size: 18pt
 *   - Center icon size: 20pt
 *
 * ## States
 * - Collapsed: Single icon showing active overlay (or stack icon when none)
 * - Expanded: Three icons arc outward (wind at -30°, rain at 0°, temperature at 30°)
 * - Selected: Active overlay glows with primary border and 20% opacity background
 *
 * ## Parameters
 * - value: Currently selected overlay type (nil for none)
 * - onValueChange: Callback when selection changes (toggle behavior: selecting same overlay deselects)
 * - availability: Which overlays are available for selection
 * - testID: Testing identifier
 */
public struct LSMinimalOverlayWidget: View {
    @Environment(\.theme) private var theme
    @State private var isExpanded = false

    private let value: LSOverlayType?
    private let onValueChange: (LSOverlayType?) -> Void
    private let availability: LSOverlayAvailability
    private let testID: String?

    /// Radial positions (angles in degrees)
    private let radialPositions: [(angle: Double, overlay: LSOverlayType)] = [
        (-30, .wind),
        (0, .rain),
        (30, .temperature),
    ]

    private let radialRadius: CGFloat = 36

    public init(
        value: LSOverlayType?,
        onValueChange: @escaping (LSOverlayType?) -> Void,
        availability: LSOverlayAvailability,
        testID: String? = nil
    ) {
        self.value = value
        self.onValueChange = onValueChange
        self.availability = availability
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        ZStack {
            // Radial icons (expanded state)
            ForEach(radialPositions, id: \.overlay) { position in
                radialIcon(for: position.overlay, at: position.angle)
            }

            // Center toggle button
            centerButton
        }
        .frame(width: 120, height: 120)
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Weather overlays")
    }

    // MARK: - Center Button

    private var centerButton: some View {
        Button(action: {
            withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                isExpanded.toggle()
            }
        }) {
            ZStack {
                // Active indicator ring
                if value != nil {
                    Circle()
                        .stroke(theme.colors.primary.default, lineWidth: 2)
                        .frame(width: 48, height: 48)
                }

                // Center button background
                Circle()
                    .fill(theme.colors.surfaceVariant.default)
                    .frame(width: 40, height: 40)
                    .overlay(
                        Circle()
                            .stroke(
                                value != nil ? theme.colors.primary.default : theme.colors.border.default,
                                lineWidth: 1.5
                            )
                    )

                // Center icon
                Image(systemName: currentIcon)
                    .font(.system(size: 20, weight: .medium))
                    .foregroundStyle(
                        value != nil ? theme.colors.primary.default : theme.colors.onSurface.default
                    )
            }
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Weather overlays")
        .accessibilityHint(isExpanded ? "Double tap to collapse" : "Double tap to expand")
    }

    // MARK: - Radial Icon

    private func radialIcon(for overlayType: LSOverlayType, at angle: Double) -> some View {
        let isAvailable = availability.isAvailable(overlayType)
        let isSelected = value == overlayType

        let radians = angle * .pi / 180
        let xOffset = CGFloat(sin(radians)) * radialRadius
        let yOffset = CGFloat(-cos(radians)) * radialRadius

        return Button(action: {
            guard isAvailable else { return }

            // Toggle behavior: selecting same overlay deselects it
            let newValue: LSOverlayType? = if value == overlayType {
                nil
            } else {
                overlayType
            }

            withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                onValueChange(newValue)
                isExpanded = false
            }
        }) {
            ZStack {
                // Background
                Circle()
                    .fill(
                        isSelected
                            ? theme.colors.primary.default.opacity(0.2)
                            : theme.colors.surfaceVariant.default
                    )
                    .frame(width: 36, height: 36)

                // Border
                Circle()
                    .stroke(
                        isSelected ? theme.colors.primary.default : theme.colors.border.default,
                        lineWidth: 1
                    )
                    .frame(width: 36, height: 36)

                // Icon
                Image(systemName: overlayType.icon)
                    .font(.system(size: 18, weight: .medium))
                    .foregroundStyle(
                        isSelected
                            ? theme.colors.primary.default
                            : theme.colors.onSurface.muted
                    )
            }
        }
        .buttonStyle(.plain)
        .offset(x: isExpanded ? xOffset : 0, y: isExpanded ? yOffset : 0)
        .opacity(isExpanded ? (isAvailable ? 1.0 : 0.4) : 0)
        .scaleEffect(isExpanded ? 1.0 : 0)
        .animation(.easeInOut(duration: 0.2), value: isExpanded)
        .disabled(!isAvailable)
        .accessibilityLabel(overlayType.accessibilityLabel)
        .accessibilityValue(isSelected ? "Selected" : "Not selected")
        .accessibilityAddTraits(isSelected ? [.isSelected, .isButton] : .isButton)
    }

    // MARK: - Computed Properties

    /// Returns the current icon to display in the center button
    private var currentIcon: String {
        guard let value else {
            return "square.stack.3d.up" // Stack icon when nothing selected
        }
        return value.icon
    }
}

// MARK: - Preview

#Preview("MinimalOverlayWidget - No selection") {
    struct PreviewWrapper: View {
        @State private var selectedValue: LSOverlayType? = nil

        var body: some View {
            LSMinimalOverlayWidget(
                value: selectedValue,
                onValueChange: { newValue in
                    selectedValue = newValue
                },
                availability: LSOverlayAvailability(
                    wind: true,
                    rain: true,
                    temperature: true
                )
            )
        }
    }

    return PreviewWrapper()
        .padding()
        .laneShadowTheme()
}

#Preview("MinimalOverlayWidget - Wind selected") {
    struct PreviewWrapper: View {
        @State private var selectedValue: LSOverlayType? = .wind

        var body: some View {
            LSMinimalOverlayWidget(
                value: selectedValue,
                onValueChange: { newValue in
                    selectedValue = newValue
                },
                availability: LSOverlayAvailability(
                    wind: true,
                    rain: true,
                    temperature: true
                )
            )
        }
    }

    return PreviewWrapper()
        .padding()
        .laneShadowTheme()
}

#Preview("MinimalOverlayWidget - Partial availability") {
    struct PreviewWrapper: View {
        @State private var selectedValue: LSOverlayType? = nil

        var body: some View {
            LSMinimalOverlayWidget(
                value: selectedValue,
                onValueChange: { newValue in
                    selectedValue = newValue
                },
                availability: LSOverlayAvailability(
                    wind: true,
                    rain: false, // Not available
                    temperature: true
                )
            )
        }
    }

    return PreviewWrapper()
        .padding()
        .laneShadowTheme()
}

#Preview("MinimalOverlayWidget - Dark theme") {
    struct PreviewWrapper: View {
        @State private var selectedValue: LSOverlayType? = .rain

        var body: some View {
            LSMinimalOverlayWidget(
                value: selectedValue,
                onValueChange: { newValue in
                    selectedValue = newValue
                },
                availability: LSOverlayAvailability(
                    wind: true,
                    rain: true,
                    temperature: true
                )
            )
        }
    }

    return PreviewWrapper()
        .padding()
        .laneShadowTheme()
        .preferredColorScheme(.dark)
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/MinimalOverlayWidget.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.components.atoms.IconSymbol
import kotlin.math.cos
import kotlin.math.sin

/**
 * Overlay type enum
 *
 * Following RN wrapper API from react-native/components/map/minimal-overlay-widget.tsx
 */
enum class OverlayType {
    WIND,
    RAIN,
    TEMPERATURE,
}

/**
 * Overlay availability data class
 *
 * Following RN wrapper API from react-native/components/map/minimal-overlay-widget.tsx
 */
data class OverlayAvailability(
    val wind: Boolean,
    val rain: Boolean,
    val temperature: Boolean,
)

/**
 * Minimal Overlay Widget molecule component
 *
 * A compact, single-icon weather overlay control that expands into a radial menu.
 * Inspired by motorcycle instrument dials and compass navigation.
 *
 * States:
 * - Collapsed: Single icon showing active overlay (or stack icon when none)
 * - Expanded: Three icons arc outward (wind, rain, temp)
 * - Selected: Active overlay glows with copper accent
 *
 * Following RN wrapper API from react-native/components/map/minimal-overlay-widget.tsx
 *
 * @param value Current selected overlay type (null for none)
 * @param onValueChange Callback when overlay selection changes
 * @param availability Which overlays have data available
 * @param testID Test ID for UI testing
 */
@Composable
fun MinimalOverlayWidget(
    value: OverlayType?,
    onValueChange: (OverlayType?) -> Unit,
    availability: OverlayAvailability,
    testID: String = "overlay-widget",
) {
    val theme = LocalLaneShadowTheme.current
    var expanded by remember { mutableStateOf(false) }

    // Icon configurations
    val overlayConfig = remember {
        mapOf(
            OverlayType.WIND to "weather-windy",
            OverlayType.RAIN to "water-outline",
            OverlayType.TEMPERATURE to "thermometer",
        )
    }

    // Radial positions (angles in degrees, radius in dp)
    data class RadialPosition(val angle: Double, val overlay: OverlayType)

    val radialPositions = remember {
        listOf(
            RadialPosition(-30.0, OverlayType.WIND),
            RadialPosition(0.0, OverlayType.RAIN),
            RadialPosition(30.0, OverlayType.TEMPERATURE),
        )
    }

    val radius = 36.dp

    // Get current icon to display
    val currentIcon = remember(value) {
        if (value != null) {
            overlayConfig[value] ?: "layers"
        } else {
            "layers"
        }
    }

    // Animation for expand/collapse
    val expandedScale: Float by animateFloatAsState(
        targetValue = if (expanded) 1f else 0f,
        animationSpec = spring(
            dampingRatio = 0.7f,
            stiffness = 150f,
        ),
        label = "expanded_scale",
    )

    Box(
        modifier = Modifier
            .size(120.dp)
            .testTag(testID),
        contentAlignment = Alignment.Center,
    ) {
        // Expanded radial icons
        radialPositions.forEach { position ->
            val isAvailable = when (position.overlay) {
                OverlayType.WIND -> availability.wind
                OverlayType.RAIN -> availability.rain
                OverlayType.TEMPERATURE -> availability.temperature
            }
            val isActive = value == position.overlay
            val iconName = overlayConfig[position.overlay] ?: "layers"

            // Calculate radial position
            val radians = Math.toRadians(position.angle)
            val offsetX = (sin(radians) * radius.value * expandedScale).dp
            val offsetY = (-cos(radians) * radius.value * expandedScale).dp

            // Animate opacity and scale
            val iconAlpha: Float by animateFloatAsState(
                targetValue = if (expanded) {
                    if (isAvailable) 1f else 0.3f
                } else {
                    0f
                },
                animationSpec = tween(durationMillis = 150),
                label = "icon_alpha_${position.overlay.name}",
            )

            val iconScale: Float by animateFloatAsState(
                targetValue = if (expanded) 1f else 0f,
                animationSpec = tween(durationMillis = 200),
                label = "icon_scale_${position.overlay.name}",
            )

            val iconSize: Dp by animateDpAsState(
                targetValue = if (expanded) 36.dp else 0.dp,
                animationSpec = tween(durationMillis = 200),
                label = "icon_size_${position.overlay.name}",
            )

            Box(
                modifier = Modifier
                    .offset(x = offsetX, y = offsetY)
                    .size(iconSize)
                    .scale(iconScale)
                    .alpha(iconAlpha),
                contentAlignment = Alignment.Center,
            ) {
                val backgroundColor = if (isActive) {
                    theme.colors.primary.default.copy(alpha = 0.2f)
                } else {
                    theme.colors.surfaceVariant.default
                }
                val borderColor = if (isActive) {
                    theme.colors.primary.default
                } else {
                    theme.colors.border.default
                }
                val iconColor = if (isActive) {
                    theme.colors.primary.default
                } else {
                    theme.colors.onSurface.default.copy(alpha = 0.6f)
                }
                val buttonAlpha = if (isAvailable) 1f else 0.4f

                Box(
                    modifier = Modifier
                        .size(36.dp)
                        .alpha(buttonAlpha)
                        .background(backgroundColor, CircleShape)
                        .border(BorderStroke(1.dp, borderColor), CircleShape)
                        .clickable(
                            enabled = isAvailable,
                            onClick = {
                                val newValue = if (value == position.overlay) null else position.overlay
                                onValueChange(newValue)
                                expanded = false
                            },
                        )
                        .semantics {
                            contentDescription = "${position.overlay.name.lowercase()} overlay"
                        },
                    contentAlignment = Alignment.Center,
                ) {
                    IconSymbol(
                        name = iconName,
                        size = 18.dp,
                        color = iconColor,
                    )
                }
            }
        }

        // Center toggle button
        val centerButtonSize: Dp by animateDpAsState(
            targetValue = 40.dp,
            animationSpec = tween(durationMillis = 200),
            label = "center_button_size",
        )

        val centerBackgroundColor = theme.colors.surfaceVariant.default
        val centerBorderColor = if (value != null) {
            theme.colors.primary.default
        } else {
            theme.colors.border.default
        }
        val centerIconColor = if (value != null) {
            theme.colors.primary.default
        } else {
            theme.colors.onSurface.default
        }

        Box(
            modifier = Modifier
                .size(centerButtonSize)
                .background(centerBackgroundColor, CircleShape)
                .border(BorderStroke(1.5.dp, centerBorderColor), CircleShape)
                .clickable(
                    onClick = { expanded = !expanded },
                )
                .semantics {
                    contentDescription = "Weather overlays"
                }
                .testTag("${testID}-center"),
            contentAlignment = Alignment.Center,
        ) {
            // Active indicator ring
            if (value != null) {
                Box(
                    modifier = Modifier
                        .size(48.dp)
                        .offset(x = (-4).dp, y = (-4).dp)
                        .background(Color.Transparent, CircleShape)
                        .border(BorderStroke(2.dp, theme.colors.primary.default), CircleShape),
                )
            }

            IconSymbol(
                name = currentIcon,
                size = 20.dp,
                color = centerIconColor,
            )
        }
    }
}
